feat: send announcement with push notification

fanlan1210 opened the pull request:
如題，使大會公告功能可以在發送時一併進行推播通知，該推播通知可以視發布者需求，在每次公告時選擇是否開啟

fanlan1210 committed:
feat: send announcement with push notification

fanlan1210 commented:
因為沒有測試環境可實驗，再請 @yoyo930021 有空時協助 review 看看程式碼的可執行性🙏

yoyo930021 commented on src/views/Announcement.vue:
```diff
@@ -92,8 +95,24 @@ export default {
+          });
+        if( this.withPushNotification ) {
+          oneSignal.createNotificationWithTagFilter({
+            target: this.role === this.options[0] ? 'all' : this.role,
```
這段好像不是我寫的，剛剛研究了一下， `this.role` 應該會是一個陣列。   
https://github.com/CCIP-App/CCIP-Admin-Bueno/blob/0c42df6edf61173bb0f35d23a1352ea2e100158c/src/module/apiClient.js#L52
https://github.com/CCIP-App/CCIP-Admin-Bueno/blob/0c42df6edf61173bb0f35d23a1352ea2e100158c/src/views/Announcement.vue#L123
可惡，想改寫成 TS 了 XDDDDD

Haraguroicha replied:
因為那段是我寫的🙃
https://github.com/CCIP-App/CCIP-Admin-Bueno/commit/b2a6509173b1d139dae0c25cbbc7846b4d823d36

fanlan1210 replied:
了解，我再來怎麼調整！
逐漸理解為什麼要有 JS doc 跟 TS XDD

fanlan1210 replied:
在 [0afc042](https://github.com/CCIP-App/CCIP-Admin-Bueno/pull/49/commits/0afc042eba766cebb1a90c4b5c2f559716772f44) 和 [3c79c7d](https://github.com/CCIP-App/CCIP-Admin-Bueno/pull/49/commits/3c79c7d58836e7cf3aef29d79d71b46af72e0010) 已經將其轉換為推播通知 API 用的字串格式，比對時也是先轉換成型態再比對，就再請 @yoyo930021 看看還有沒有其他問題了。

fanlan1210 committed:
fix: correct the announcement role type check

fanlan1210 committed:
fix: converted this.newAnnounce.role to string for push notification

yoyo930021 commented on src/views/Announcement.vue:
```diff
@@ -94,6 +97,22 @@ export default {
           })
+        if (this.withPushNotification) {
+          oneSignal.createNotificationWithTagFilter({
+            target: this.newAnnounce.role.toString() === this.options[0].toString() ? 'all' : this.newAnnounce.role[0],
```
這個判斷還是有問題，根據我實際跑起來，當選擇全體的時候，`this.newAnnounce.role` 會是一個包含所有選項值的陣列。
你可以實體跑起來看看，觀察它實際的值。

fanlan1210 replied:
感謝 @yoyo930021 指出問題，
我發現是我在修正時沒正確取到 options 裡的 value，造成將 newAnnounce.role 轉換為 oneSignal 接受的 target 型態判斷時發生錯誤。
目前已經花了點時間模擬實際情況，且測試過後可以正確帶 target 值給 oneSignal，再麻煩 IU review 了，大感謝> <

fanlan1210 committed:
fix: role comparison in announcement push notification

The change involves correcting the role comparison logic for push
notifications by properly accessing the `value` property.

yoyo930021 reviewed (changes requested):
現在程式碼執行上應該沒問題了，我只是在想，現在的錯誤訊息分不出來：
- 發公告失敗
- 發推播失敗
- 兩者都失敗

fanlan1210 commented:
> 現在程式碼執行上應該沒問題了，我只是在想，現在的錯誤訊息分不出來：
> 
> * 發公告失敗
> * 發推播失敗
> * 兩者都失敗

好，那我來想想看怎麽設計，可以讓使用者知道失敗的具體狀況
